### PR TITLE
feat(activerecord): sanitization.rb privates (5/5)

### PR DESCRIPTION
**Pull request by deanmarano**
## Summary

Implement the 5 private helper methods in \`sanitization.ts\` to close the private-API gap for Rails' \`sanitization.rb\`:

- \`replaceBindVariables\` — replaces \`?\` placeholders with quoted values
- \`replaceBindVariable\` — quotes a single bind variable
- \`replaceNamedBindVariables\` — replaces \`:name\` style bind variables with support for PostgreSQL type casts
- \`quoteBoundValue\` — quotes values, handling arrays and \`id_for_database\` protocol
- \`raiseIfBindArityMismatch\` — validates bind variable count matches placeholders

## Implementation details

The implementation mirrors the Rails source from \`activerecord/lib/active_record/sanitization.rb\`:
- Handles both positional (\`?\`) and named (\`:name\`) bind variables
- Supports PostgreSQL type casts (\`::type\`) and escaped colons in named bind patterns
- Properly quotes arrays as comma-separated values
- Respects \`id_for_database\` protocol for custom value types
- Validates bind variable arity before processing

## Test plan

- Added comprehensive tests in \`sanitize.test.ts\` covering:
  - Named bind variables with various types (strings, numbers, null, booleans)
  - Single quote escaping in values
  - PostgreSQL type cast handling
  - Array and empty array handling
  - Missing bind variable errors
  - Arity mismatch errors

## Verification

- \`pnpm api:compare --package activerecord --privates-only\` shows: \`sanitization.rb 5 0 5 100%\` (was 0/5)
- All existing tests pass
- New test suite validates private method behavior

See: docs/private-api-parity-100-plan.md — Tier 7, smallest target.

### CODE REVIEW

**Comment by deanmarano** [comment]
CI failure (PostgreSQL/MariaDB/SQLite Tests, run 25055753998): the failing test is **`sanitize.test.ts > handles escaped colons in named bind variable patterns`** — part of this PR's own diff, not a flake.

Bug in `replaceNamedBindVariables`: escaped colons (`\:`) are not being unescaped after substitution.

```
Expected to contain: "10:00"
Got:                 "time_value = '14:30' AND literal_colon = '10\:00'"
```

Rails strips the leading backslash from `\:` after substitution. Fix the implementation to do the same. Blocking merge.

**Comment by deanmarano** [comment]
## Review: sanitization.rb privates

### Criteria Assessment

**(1) Implementation matches Rails ✓**
- All 5 private methods (replaceBindVariables, replaceBindVariable, replaceNamedBindVariables, quoteBoundValue, raiseIfBindArityMismatch) match Rails semantics exactly
- Regex patterns, error handling, and value quoting are correct
- Acceptable TS adaptation: uses module-level `quote()` instead of `connection.quote()`

**(2) No stubs ✓**
- All NotImplementedError placeholders replaced with real implementations
- No throw statements or unimplemented methods remain

**(3) api:compare delta ✓**
- Origin/main: sanitization.rb 0/5 (0%)
- PR branch: sanitization.rb 5/5 (100%)
- Delta: +5 methods ✓

### Blocking Issue

**Test failure: "handles escaped colons in named bind variable patterns"**

Line 437-441 of sanitize.test.ts:
```ts
const result = Post.sanitizeSqlArray("time_value = :time AND literal_colon = '10\\:00'", {
  time: "14:30",
});
expect(result).toContain("10:00");
```

**Problem:** The test input `'10\:00'` doesn't trigger the escaped-colon handling because the regex `/([:\\]?):([a-zA-Z]\w*)/` only matches when `:` is followed by `[a-zA-Z]` (a valid identifier start). Since `0` is not a letter, `\:00` doesn't match, so the backslash isn't removed. Rails behaves the same way.

**Fix:** Update the test to use a valid identifier after the escaped colon:
```ts
const result = Post.sanitizeSqlArray("time_value = :time AND pattern = 'HH\\:MI\\:SS'", {
  time: "14:30",
});
expect(result).toContain("HH:MI:SS");
```

This matches the Rails documentation example (line 151 of sanitization.rb) which shows `HH12\:MI\:SS` → `HH12:MI:SS`.

Once the test is fixed, this PR is ready to merge. ✓

**Comment by deanmarano** [comment]
Fixed the failing test: updated 'handles escaped colons in named bind variable patterns' to use Rails' documented example format (YYYY/MM/DD HH12\:MI\:SS) where backslashes escape colons before letters. The regex /([:\]?):([a-zA-Z]\w*)/ only matches colons followed by letters, so the test was incorrect before. Commit: 5f0d5960

**Comment by deanmarano** [comment]
CI is green! All checks passed:
- Build & Type Check: pass
- DX Type Tests: pass
- Lint: pass
- Prettier: pass
- Rails API/Test Comparison: pass
- Unit Tests: pass
- Virtualized DX Type Tests: pass
- SQLite Tests: pass
- MariaDB Tests: pass
- PostgreSQL Tests: pass
- Website: pass

**Comment by deanmarano** [comment]
## Re-review Complete ✓

All verification criteria passed:

### 1. Implementation vs Rails Source
- **replaceBindVariables**: Matches Rails line-for-line — arity check + gsub with callback ✓
- **replaceBindVariable**: Delegates to quoteBoundValue ✓
- **replaceNamedBindVariables**: Regex  correctly handles PostgreSQL casts, escaped colons, and named variables ✓
- **quoteBoundValue**: Handles id_for_database protocol, arrays (including empty), and primitives ✓
- **raiseIfBindArityMismatch**: Validates expected === provided ✓

### 2. No Stubs
All five NotImplementedError stubs replaced with working implementations. No remaining stubs in changed files. ✓

### 3. api:compare Results
- **Before**: sanitization.rb 0/5
- **After**: sanitization.rb 5/5 (100%) ✓

### 4. CI Status
All 13 primary test jobs green:
- Build & Type Check ✓
- SQLite/MariaDB/PostgreSQL Tests ✓
- DX Type Tests & Virtualized DX Type Tests ✓
- Rails API/Test Comparison ✓
- Lint, Prettier ✓

### 5. Test Fix Sensibility (Commit 5f0d5960)
The escaped colon test fix is correct:
- Old test: `'10\:00'` — colon followed by digit (0), does NOT match regex requirement `[a-zA-Z]`
- New test: `'HH12\:MI\:SS'` — escaped colons before letters (M, I, S), exercises the escaping code path ✓
- Matches Rails' documented example exactly ✓

**PR Size**: 249 LOC (217 additions + 32 deletions), well under 300 LOC ceiling ✓

**Ready to merge.**

**Comment by deanmarano** [comment]
Addressed Copilot review #1 (commit f2bc0927):

**Comment 2 (named-bind detection too broad)** — Fixed. Added `isPlainHash()` (Object.prototype or null proto) and gate the named-bind branch on it. Date/Arel nodes/model instances no longer route to named-bind.

**Comment 3 (`in` walks prototype chain)** — Fixed. Switched to `Object.prototype.hasOwnProperty.call(bindVars, name)`, matching Ruby `Hash#has_key?`.

**Comment 4 (Relation/toSql doc lie)** — Fixed. Removed the misleading doc line. Rails' `quote_bound_value` doesn't handle Relations either — Relation-as-bind inlining happens in PredicateBuilder for hash conditions (`where(x: relation)`), not here.

**Bonus (idForDatabase camelCase)** — While in the same code path, switched `id_for_database` → `idForDatabase` to match the actual method name on PrimaryKey/CompositePrimaryKey records (per CLAUDE.md camelCase rule).

**Comment 1 (extra binds with no placeholders should raise arity mismatch)** — Declining. Rails' `sanitize_sql_array` does NOT raise on extra binds when neither `:name` nor `?` is present:

```ruby
def sanitize_sql_array(ary)
  statement, *values = ary
  if values.first.is_a?(Hash) && /:\w+/.match?(statement)
    replace_named_bind_variables(statement, values.first)
  elsif statement.include?("?")
    replace_bind_variables(statement, values)
  elsif statement.blank?
    statement
  else
    statement % values.collect { |value| connection.quote_string(value.to_s) }
  end
end
```

When there are no placeholders, Rails uses Ruby's `%` string format — extra values are silently ignored, not raised. The TS impl is permissive in the same direction (returns the statement unchanged) and could be extended to do the `%` formatting later, but routing through `replaceBindVariables` (which raises arity) would deviate from Rails.

**Comment by deanmarano** [comment]
# Deep Rails-Fidelity Review of PR #947

After comparing the TypeScript implementation line-by-line against Rails activerecord/lib/active_record/sanitization.rb (commit context), I found the following Rails-fidelity gaps:

## 1. BLOCKING: Missing `%s` format string support (Severity: Blocking)

**Rails behavior** (sanitization.rb lines 177-179):
```ruby
else
  with_connection do |c|
    statement % values.collect { |value| c.quote_string(value.to_s) }
  end
```

**TS behavior** (sanitization.ts lines 29-33):
```typescript
if (statement === "") {
  return statement;
}

return statement;  // Returns statement unchanged
```

**Gap**: When a SQL statement contains `%s` placeholders (no `?` and no `:name` binds), Rails applies Ruby's `%` operator to interpolate the values. The TS implementation returns the template unchanged.

**Example**:
- Rails: `sanitize_sql_array(["name='%s' and group_id='%s'", "foo'bar", 4])` → `"name='foo''bar' and group_id='4'"`
- TS: Same input → `"name='%s' and group_id='%s'"` (unchanged)

**Rails reference**: sanitization.rb line 154-155 documents this case explicitly.

**Fix**: Implement the `%s` interpolation logic when neither `?` nor `:` placeholders are found and values are present:
```typescript
if (statement === "") {
  return statement;
}

if (binds.length > 0) {
  // %s format string case: interpolate quoted values
  const formatted = binds.map((v) => quote(String(v)));
  return statement.replace(/%s/g, () => formatted.shift() || "%s");
}

return statement;
```

---

## 2. BLOCKING: Array/Range handling mismatch (Severity: Blocking)

**Rails behavior** (sanitization.rb line 234):
```ruby
if value.respond_to?(:map) && !value.acts_like?(:string)
  values = value.map { |v| v.respond_to?(:id_for_database) ? v.id_for_database : v }
  # ... handle as array
```

**TS behavior** (sanitization.ts line 292):
```typescript
if (Array.isArray(value)) {
  // ... handle as array
}
```

**Gap**: Rails uses duck-typing (`respond_to?(:map)`) to handle any enumerable type (Array, Range, Set, etc.). The TS implementation only handles native JS Arrays, not Range-like objects or other iterables.

**Impact**: A user passing a Range object as a bind value would be treated as a scalar and passed to `quote()` instead of being expanded to comma-separated values.

**Example** (hypothetical but valid in Rails):
- Rails: `sanitize_sql_array(["id IN (?)", (1..3)])` → `"id IN (1,2,3)"`
- TS: Same input → `"id IN ('Range...')"` or similar (depends on quote behavior)

**Rails reference**: sanitization.rb lines 234–245.

**Fix**: Implement a helper to detect array-like objects:
```typescript
function isArrayLike(value: unknown): boolean {
  if (Array.isArray(value)) return true;
  if (typeof value !== "object" || value === null) return false;
  // Check if it has a .map method and is not a string-like object
  return typeof (value as any).map === "function" && 
         typeof (value as any).acts_like_string !== "function";
}
```

Then replace `if (Array.isArray(value))` with `if (isArrayLike(value))`.

---

## 3. BLOCKING: cast_bound_value not applied to values (Severity: Medium)

**Rails behavior** (sanitization.rb lines 237, 239):
```ruby
connection.quote(connection.cast_bound_value(nil))
values.map! { |v| connection.quote(connection.cast_bound_value(v)) }.join(",")
```

**TS behavior** (sanitization.ts line 294, 296):
```typescript
return quote(null);
return value.map((v) => (hasIdForDatabase(v) ? quote(v.idForDatabase()) : quote(v))).join(",");
```

**Gap**: Rails always calls `connection.cast_bound_value()` before quoting each value, even for arrays. The TS implementation skips this step.

**Current impact**: Minimal (since `castBoundValue` is currently a no-op in TS), but when cast_bound_value is enhanced to handle type-specific casting (e.g., JSON, binary), the TS code will diverge.

**Rails reference**: sanitization.rb lines 233–245.

**Fix**: Apply `castBoundValue` before quoting:
```typescript
function quoteBoundValue(value: unknown): string {
  if (hasIdForDatabase(value)) {
    return quote(castBoundValue(value.idForDatabase()));
  }

  if (isArrayLike(value)) {
    if (value.length === 0) {
      return quote(castBoundValue(null));
    }
    return value.map((v) => {
      if (hasIdForDatabase(v)) {
        return quote(castBoundValue(v.idForDatabase()));
      }
      return quote(castBoundValue(v));
    }).join(",");
  }

  return quote(castBoundValue(value));
}
```

---

## 4. NICE-TO-HAVE: Blank string handling (Severity: Minor)

**Rails behavior** (sanitization.rb line 174):
```ruby
elsif statement.blank?  # checks for nil, empty string, and whitespace-only strings
```

**TS behavior** (sanitization.ts line 29):
```typescript
if (statement === "")  // only checks for exactly empty string
```

**Gap**: Rails' `blank?` returns true for nil, empty strings, and strings that contain only whitespace. The TS check only handles exactly empty strings.

**Impact**: A statement like `"   "` (spaces only) would be treated as having no placeholders and return unchanged in Rails but also return unchanged in TS (since it falls through), so behavior is identical. However, the intent differs.

**Rails reference**: sanitization.rb line 174.

**Fix** (optional, minimal impact):
```typescript
if (!statement || statement.trim() === "") {
  return statement;
}
```

---

## 5. NICE-TO-HAVE: acts_like?(:string) exclusion in array detection

**Rails behavior** (sanitization.rb line 234):
```ruby
if value.respond_to?(:map) && !value.acts_like?(:string)
```

**TS behavior**: N/A (not checking `acts_like?(:string)`)

**Gap**: Rails explicitly excludes string-like objects (those implementing `acts_like_string?`) from the array branch. This prevents accidentally treating a string-like object's `.map` method as an indicator of iterability.

**Impact**: Minimal in typical usage, but if a custom class implements both `.map` and `acts_like_string?`, Rails would quote it as a scalar while TS would iterate it.

**Rails reference**: sanitization.rb line 234.

**Fix**: Add a check when implementing array-like detection:
```typescript
function isArrayLike(value: unknown): boolean {
  if (Array.isArray(value)) return true;
  if (typeof value !== "object" || value === null) return false;
  const hasMap = typeof (value as any).map === "function";
  const actsLikeString = typeof (value as any).acts_like_string === "function";
  return hasMap && !actsLikeString;
}
```

---

## Summary

| Gap | Severity | Status | Affects |
|-----|----------|--------|---------|
| `%s` format string support | Blocking | Not implemented | Direct `sanitizeSqlArray` calls with format strings |
| Range/enumerable handling | Blocking | Not implemented | Any Range/Set passed as bind value |
| `cast_bound_value` application | Medium | Partially implemented | Future type-specific casting |
| Blank string check | Minor | Functionally same | Edge case only |
| `acts_like?(:string)` check | Nice-to-have | Not implemented | Custom string-like objects with `.map` |

---

## Tests

All existing tests pass. New tests added in the PR cover:
- Named bind variables (strings, numbers, mixed types)
- Escaped colons and PostgreSQL type casts
- Empty and non-empty arrays
- Missing bind variable errors
- Arity mismatch errors

**Recommended**: Add tests for the blocking gaps:
1. `%s` format string interpolation
2. Range objects as bind values

**Comment by deanmarano** [comment]
Addressed Copilot review #2:

**Comment 1 (regex doesn't match backslash prefix)** — Declining; the regex is correct. Verified with Node:

```
$ node -e 'const m = "x\\:foo".match(/([:\\\\]?):([a-zA-Z]\\w*)/); console.log(JSON.stringify(m));'
["\\\\:foo","\\\\","foo"]
```

In a JS regex literal, `\\` is a single literal backslash, so the character class `[:\\]` matches `:` or `\`. The escaped-colon branch (`prefix === "\\"`) does fire.

This is also exercised by the passing test `handles escaped colons in named bind variable patterns` (commit 5f0d5960), which asserts that `'HH12\\:MI\\:SS'` → `'HH12:MI:SS'`. If the branch were dead, that test would fail; it passes in CI.

(Copilot may have parsed the regex as `/([:\\]?):/` literally rather than treating `\\\\` in source as the JS escape for a single backslash.)

File: packages/activerecord/src/sanitization.ts [hunk 1]
```diff
@@ -226,30 +240,134 @@ export const ClassMethods = {
...
+  result = result.replace(/\?/g, () => replaceBindVariable(bound.shift()));
+  return result;
+}
+
+/**
+ * Quote a single bind variable value.
+ * Mirrors: ActiveRecord::Sanitization::ClassMethods#replace_bind_variable
+ */
+function replaceBindVariable(value: unknown): string {
+  return quoteBoundValue(value);
```
[the remark this answers is not in the document]
Comment (deanmarano) [reply packages/activerecord/src/sanitization.ts:265]: Already correct: the regex is `([:\\]?)` (a character class containing `:` and `\\`), which does match a literal backslash prefix. The `prefix === "\\\\"` branch in the handler correctly strips the escape and returns the literal colon. This was verified against Rails' `replace_named_bind_variables` behavior.

**Comment by deanmarano** [comment]
Three Rails-fidelity gaps re-applied and tests un-skipped:

1. **%s format string support** — Added sanitizeSqlArray branch that detects '%s' placeholders (when neither '?' nor ':name' exist) and interpolates values via quoteString(), matching Rails sanitization.rb:154-166
2. **Set/iterable support** — Added isEnumerable() helper that detects Array, Set, and other iterables. Updated quoteBoundValue() to convert Sets to Arrays via Array.from() and process element-by-element
3. **castBoundValue routing** — Updated all paths in quoteBoundValue() (scalar, idForDatabase, arrays/Sets) to call adapter.castBoundValue() before quote(), matching Rails quote_bound_value delegation

Tests:
- All 41 tests pass (3 previously skipped now active)
- Commit: 235d2fc2
- PR: MERGEABLE (mergeStateStatus: CLEAN)
- diff: 55 additions/11 deletions (under 300 LOC)

**Comment by deanmarano** [comment]
Addressed Copilot review #3:

All 3 comments are about `it.skip(...)` calls for `%s format string`, `handles Sets`, and `handles empty Sets` lacking skip-reason comments. **These tests are no longer skipped** as of commit 235d2fc2 — the underlying implementations (`%s` format substitution, Set/iterable support in `quoteBoundValue`, empty-Set → NULL) were re-applied after the rebase, and the tests are live (`it(...)`) and passing.

Verified on the current sha:
```
$ grep -n 'sanitize sql array handles %s\|handles Sets as\|handles empty Sets' packages/activerecord/src/sanitize.test.ts
377:    it("sanitize sql array handles %s format string", () => {
488:    it("handles Sets as bind values", () => {
498:    it("handles empty Sets as bind values", () => {
```

Copilot's review was generated against an older sha (`cd6b4ee6`) where the rebase had temporarily skipped those tests. No action needed.

**Comment by deanmarano** [comment]
Addressed Copilot review #4 (commit 63274e49):

**Comment 1 (`%s` arity / null crash)** — Fixed. The `%s` branch now counts placeholders via `(statement.match(/%s/g) ?? []).length`, calls `raiseIfBindArityMismatch` on count mismatch (same path as the `?` placeholders), and coerces values with `String(value ?? '')` so null/undefined produce `''` instead of throwing on `.toString()`. Mirrors Rails `statement % values.collect { |v| connection.quote_string(v.to_s) }` with explicit arity enforcement.

**Comment 2 (`isEnumerable` too loose)** — Fixed. Now accepts `Array` and `Set` only — dropped the `Symbol.iterator in value` branch that would have expanded `Buffer`/`Map`/etc as bind collections. Updated the comment to match the behavior.

Added regression tests for arity mismatch and nullish-%s coercion (43 tests pass, 4 skipped).

**Comment by deanmarano** [comment]
Addressed Copilot review #5 (commit 376701c6): updated the comment above the `isEnumerable` check to match the actual behavior — Array and Set only, not arbitrary iterables. Rails reference (`respond_to?(:map)`/`acts_like?(:string)`) preserved for context with a note that the TS impl intentionally narrows to those two types.

**Comment by deanmarano** [comment]
Addressed Copilot review #6: the requested tests already landed in commit 63274e49 (review #4 fix):

- `sanitize sql array %s format raises on arity mismatch` at sanitize.test.ts:385 — verifies `Post.sanitizeSqlArray("name='%s' and id='%s'", 'foo')` throws `/wrong number of bind variables/`.
- `sanitize sql array %s format coerces nullish to empty string` at sanitize.test.ts:394 — verifies `Post.sanitizeSqlArray("name='%s'", null)` returns `\`name=''\``.

Both pass on the current sha (376701c6). No additional changes needed.

**Comment by deanmarano** [comment]
Website CI failure (run 25074940744 job 73465622092): runner infrastructure issue — "The self-hosted runner lost communication with the server" during the typedoc/VitePress docs build. Unrelated to this PR's diff (sanitization changes don't touch website). Stopping per instructions; recommend rerunning the Website job.
